src/Operation/Recommendation/Filters.php

```diff
@@ -0,0 +1,215 @@
+
+    /** @var boolean */
+    private $discounted;
+
+    /** @var boolean */
+    private $fresh;
+
+    /** @var array */
+    private $price;
+
+    /** @var string[] */
+    private $productIds;
+
+    /** @var float */
+    private $rating;
+
+    /** @var int */
+    private $reviews;
+
+    /** @var string */
+    private $search;
+
+    private $stock;
+    private $tag1;
+    private $tag2;
+    private $tag3;
+
+    /**
+     * @param string[] $brands
+     */
+    public function setBrands(array $brands)
+    {
+        $this->brands = array_merge($this->brands, $brands);
+    }
+
+    /**
+     * @param string[] $categories
+     */
+    public function setCategories(array $categories)
+    {
+        $this->categories = array_merge($this->categories, $categories);
+    }
+
+    /**
+     * @param string $attribute
+     * @param array $values
+     */
+    public function setCustomFields($attribute, array $values)
+    {
+        foreach ($this->customFields as $customField) {
+            if ($customField['attribute'] === $attribute) {
+                $customField['values'] = array_merge($customField['values'], $values);
+            }
+        }
+
+        $this->customFields[] = [
+            'attribute' => $attribute,
+            'values' => $values
+        ];
+    }
+
+    /**
+     * @param boolean $discounted
+     */
+    public function setDiscounted($discounted)
+    {
+        $this->discounted = $discounted;
+    }
+
+    /**
+     * @param boolean $fresh
+     */
+    public function setFresh($fresh)
+    {
+        $this->fresh = $fresh;
+    }
+
+    /**
+     * @param int $min
+     * @param int $max
+     */
+    public function setPrice($min, $max)
+    {
+        $this->price = ['min' => $min, 'max' => $max];
+    }
+
+    /**
+     * @param string[] $productIds
+     */
+    public function setProductIds(array $productIds)
+    {
+        $this->productIds = array_merge($this->productIds, $productIds);
+    }
+
+    /**
+     * @param float $rating
+     */
+    public function setRating($rating)
+    {
+        $this->rating = $rating;
+    }
+
+    /**
+     * @param int $reviews
+     */
+    public function setReviews($reviews)
+    {
+        $this->reviews = $reviews;
+    }
+
+    /**
+     * @param string $search
+     */
+    public function setSearch($search)
+    {
+        $this->search = $search;
+    }
+
+    /**
+     * @param mixed $stock
+     */
+    public function setStock($stock)
+    {
+        $this->stock = $stock;
+    }
+
+    /**
+     * @param mixed $tag1
+     */
+    public function setTag1($tag1)
+    {
+        $this->tag1 = $tag1;
+    }
+
+    /**
+     * @param mixed $tag2
+     */
+    public function setTag2($tag2)
+    {
+        $this->tag2 = $tag2;
+    }
+
+    /**
+     * @param mixed $tag3
+     */
+    public function setTag3($tag3)
+    {
+        $this->tag3 = $tag3;
+    }
+
+    /**
+     * @return array
+     * @suppress PhanTypeSuspiciousNonTraversableForeach
```

Comment:
Typehint `$this` to Traversable or extend from Traversable to get rid of this warning